sympy/sets/sets.py

```diff
@@ -352,14 +352,34 @@ def contains(self, other):
         return b
 
     def _contains(self, other):
-        raise NotImplementedError(filldedent('''
-            (%s)._contains(%s) is not defined. This method, when
-            defined, will receive a sympified object. The method
-            should return True, False, None or something that
-            expresses what must be true for the containment of that
-            object in self to be evaluated. If None is returned
-            then a generic Contains object will be returned
-            by the ``contains`` method.''' % (self, other)))
+        """Test if ``other`` is an element of the set ``self``.
+
+        This is an internal method that is expected to be overridden by
+        subclasses of ``Set`` and will be called by the public
+        :func:`Set.contains` method or the :class:`Contains` expression.
+
+        Parameters
+        ==========
+
+        other: Sympified :class:`Basic` instance
+            The object whose membership in ``self`` is to be tested.
+
+        Returns
+        =======
+
+        Symbolic :class:`Boolean` or ``None``.
+
+        A return value of ``None`` indicates that it is unknown whether
+        ``other`` is contained in ``self``. Returning ``None`` from here
+        ensures that ``self.contains(other)`` or ``Contains(self, other)`` will
+        return an unevaluated :class:`Contains` expression.
+
+        If not ``None`` then the returned value is a :class:`Boolean` that is
```

Comment:
This seems a little odd to me: I can see that `And` might be returned but I wouldn't expect it to be unevaluated if all elements are known, only if something is not known. But if not known, then shouldn't it return None? Or maybe there is something else that comes back from here?

Reply:
Here are some examples:
```python
In [4]: FiniteSet(x, y).contains(1)
Out[4]: x = 1 ∨ y = 1

In [5]: Intersection({x}, {y}).contains(1)
Out[5]: x = 1 ∧ y = 1

In [6]: (S({x}) - S({y})).contains(z)
Out[6]: x = z ∧ y ≠ z

In [7]: Interval(x, y).contains(z)
Out[7]: z ≥ x ∧ z ≤ y

In [9]: ConditionSet(x, x>0).contains(y)
Out[9]: y > 0

In [10]: (Reals**2).contains((x, y))
Out[10]: -∞ < x ∧ -∞ < y ∧ x < ∞ ∧ y < ∞
```
In each of the cases `Contains` would not evaluate but `.contains` will return a Boolean other than Contains if possible.

Reply:
It seems like this is the result of returning None from `_contains` -- at least master returns None for `FiniteSet(x, y)._contains(1)`. I can understand that. 

What the docstring is not helping with is giving direction as to when None should be returned or when an "logically equivalent" expression consistent with containment be returned. Looking through this file, I only see `Range` returning something other than true, false or None. What criteria should be used to decide whether to return None or something else? (I feel like there should be more elaboration of that in this docstring.)

Reply:
(And I do appreciate that you have already improved the docstring...I am just thinking that there should be more.)

Reply:
The `_contains` method is confused because I guess at one point it used three valued logic and then maybe later it was changed to sometimes return a symbolic Boolean. In the former case a return value of None makes sense but in the latter it does not.

The method can be called in several different situations:

1. `Contains(x, s)`
2. `s.contains(x)`
3. `x in s`

It is also called as part of three-valued-logic queries like:
https://github.com/sympy/sympy/blob/13276caf5971cf3205ff4311bc84493c3581ad2b/sympy/sets/handlers/comparison.py#L26-L34

The only effect of returning `None` here rather than some other non-true or false Boolean is just that in the `s.contains(x)` case we would get a `Contains`:
```python
In [2]: imageset(lambda n: 3*n, Integers).contains(y)
Out[2]: y ∈ {3⋅n │ n ∊ ℤ}
```
Potentially we could just say that it should return an unevaluated Contains in that case although care would be needed to avoid infinite recursion since this is called by `Contains.__new__`. The `s.contains(x)` case is also the only case that would actually make use of a result other than true, false or None.

Probably what would make the most sense is to have two separate methods. One method would just use three-valued-logic and always return True, False or None without creating any symbolic booleans. Then the other method would always return a symbolic Boolean that if not true or false could be Contains but would usually be some other kind of Boolean. Let's call these two methods `_contains` and `_contains_symbolic`. Then the only place that would actually use `_contains_symbolic` would be the `s.contains(x)` method although perhaps it could also be used by `Contains._eval_simplify` or `Contains.doit` or something. Everywhere else would use the `_contains` method converting (if necessary) True and False to true and false.

Reply:
Could `_contains` get another argument that indicates what to do when the answer is not true or false, i.e. is symbolic? e.g. `def _contains(self, other, tfn=True)` would return true, false or None. If `tfn=False` then true, false or an equivalently true (but not an unevaluated version of the input) would be returned.

Another idea: is `_contains`, by not returning None, doing the work that the calling routine should be doing? The calling routine should perhaps be responding to None and building the appropriate unevaluated Contains or symbolically equivalent expression instead of `_contains`?

This feels a little bit like the `subs`, `_subs`, `_eval_subs` construction.